robolectric/src/test/java/org/robolectric/shadows/ShadowBiometricManagerTest.java

```diff
@@ -21,22 +25,78 @@ public class ShadowBiometricManagerTest {
     biometricManager =
-            ApplicationProvider.getApplicationContext().getSystemService(BiometricManager.class);
+        ApplicationProvider.getApplicationContext().getSystemService(BiometricManager.class);
     assertThat(biometricManager).isNotNull();
   }
 
   @Test
   public void testCanAuthenticate_serviceNotConnected_canNotAuthenticate() {
-    ((ShadowBiometricManager) Shadow.extract(biometricManager)).setCanAuthenticate(false);
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.setCanAuthenticate(false);
 
     assertThat(biometricManager.canAuthenticate())
         .isEqualTo(BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE);
   }
 
   @Test
   public void testCanAuthenticate_serviceConnected_canAuthenticate() {
-    ((ShadowBiometricManager) Shadow.extract(biometricManager)).setCanAuthenticate(true);
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.setCanAuthenticate(true);
 
     assertThat(biometricManager.canAuthenticate()).isEqualTo(BiometricManager.BIOMETRIC_SUCCESS);
   }
+
+  @Test
+  public void testCanAuthenticate_serviceNotConnected_noEnrolledBiometric_biometricNotEnrolled() {
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.setEnrolledBiometrics(0);
+
+    assertThat(biometricManager.canAuthenticate())
+        .isEqualTo(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED);
+  }
+
+  @Test
+  public void testCanAuthenticate_serviceNotConnected_noHardware_biometricHwUnavailable() {
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.hasHardware(false);
+
+    assertThat(biometricManager.canAuthenticate())
+        .isEqualTo(BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE);
+  }
+
+  @Test
+  @Config(minSdk = R)
+  public void
+      testCanAuthenticate_serviceNotConnected_securityUpdateRequired_biometricErrorSecurityUpdateRequired() {
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.requireSecurityUpdate(true);
+
+    assertThat(biometricManager.canAuthenticate())
+        .isEqualTo(BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED);
+  }
+
+  @Test
+  @Config(minSdk = R)
+  public void testCanAuthenticateBiometricWeak_NoWeakButHaveStrongEnrolled_canAuthenticate() {
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.setCanAuthenticate(true);
+    shadowBiometricManager.setAuthenticatorType(BiometricManager.Authenticators.BIOMETRIC_STRONG);
+
+    assertThat(biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK))
+        .isEqualTo(BiometricManager.BIOMETRIC_SUCCESS);
+  }
+
+  @Test
+  @Config(minSdk = R)
+  public void testCanAuthenticate_ReturnsError_WhenUnsecured_CredentialAllowed_OnApi29() {
```

Comment:
Rename to `testCanAuthenticate_serviceConnected_returnsError_whenUnsecured_credentialAllowed_fromR`? It's better if we can keep code style consistent with current file existing code. And I have a question, why this method should add something like `OnApi29`? Why does not above test method `testCanAuthenticateBiometricWeak_NoWeakButHaveStrongEnrolled_canAuthenticate` have?

Reply:
I think it should be my bad, i copied the test case from here (https://github.com/androidx/androidx/blob/4d5e325e8452635286aed1101fff2ac7bfd4453e/biometric/biometric/src/test/java/androidx/biometric/BiometricManagerTest.java) but forgot to align the test case name. I can keep the naming align with other test cases